src/client/Client.cc

```diff
@@ -10055,16 +10067,63 @@ int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
     }
   }
 
+  if (onfinish) {
+     crf_iofinish = new CRF_iofinish();
+     iofinish.reset(crf_iofinish);
+
+     crf.reset(new
+       C_Read_Finisher(this, onfinish, crf_onuninline, iofinish.get(),
+                       !conf->client_debug_force_sync_read &&
+                         conf->client_oc &&
+                         (have & (CEPH_CAP_FILE_CACHE |
+                                  CEPH_CAP_FILE_LAZYIO)),
+                       have, movepos, start, f, in, f->pos, offset, size));
+
+    if (crf_onuninline != NULL)
+      crf_onuninline->CRF = crf.get();
+    crf_iofinish->CRF = crf.get();
+  }
+
   if (!conf->client_debug_force_sync_read &&
       conf->client_oc &&
       (have & (CEPH_CAP_FILE_CACHE | CEPH_CAP_FILE_LAZYIO))) {
 
     if (f->flags & O_RSYNC) {
       _flush_range(in, offset, size);
     }
-    rc = _read_async(f, offset, size, bl);
+    rc = _read_async(f, offset, size, bl, iofinish.get());
+
+    if (onfinish) {
+      // handle async caller (onfinish != nullptr), we can now safely release
+      // all the managed pointers, but we might need to do something with
+      // iofinisher.
+      Context *iof = iofinish.release();
+      onuninline.release();
+      crf.release();
+
+      if (rc < 0)
+        iof->complete(rc);
+
+      // allow caller to wait on onfinish...
+      return 0;
+    }
+
     if (rc < 0)
       goto done;
+  } else if (onfinish) {
+    // handle _sync_read asynchronously...
```

Comment:
I added comments to hopefully make this make more sense... There is a method, Client::_read_sync which is a code path the async (non-blocking) path still needs to follow, but it can't actually call _read_sync() to do it, thus the C_Read_Sync_Async class to manage the asynchronous non-blocking execution of that code path...

Maybe there's a better way to name it, I just wanted to help tie it into the blocking code path it emulates...

Reply:
thanks for the comments. I'm able to understand this better.